library/common/http/dispatcher.cc

```diff
@@ -96,131 +95,109 @@ void Dispatcher::DirectStreamCallbacks::encodeHeaders(const ResponseHeaderMap& h
 
   ENVOY_LOG(debug, "[S{}] intercepted local response", direct_stream_.stream_handle_);
   if (end_stream) {
-    // The local stream may or may not have completed.
-    // If the local is not closed envoy will fire the reset for us.
-    // @see Dispatcher::DirectStreamCallbacks::closeRemote.
-    // Otherwise fire the reset from here.
-    if (direct_stream_.local_closed_) {
-      onReset();
-    }
+    // If the stream is not locally closed, the HCM will fire resetStream subsequently. However, we will see
+    // that the stream is already gone from the streams_ map, and direct Envoy to runResetCallbacks() without
+    // redundantly calling on onReset().
+    // @see Dispatcher::DirectStreamCallbacks::resetStream.
+    onReset();
```

Comment:
I would somehow move this call back out to the calling function (either via return code or something else) so that all of the reset/closure logic is in one place.

Reply:
Sure - in fact, your suggestion led to me to realize I can simplify this even further, by checking success prior to this special handling.